[FastFile] add optional path parameter to #parse method

samnung opened the pull request:
# parse method now has optional parameter path to parsing file, this is helpful when printing backtrace to console.

I also found problem with binding, some testing FastFile uses this new variable (actually there was method with same name,
but local variable has higher priority in this case) and causes to rewrite whole file.
So I add method #get_binding to resolve this issue.

Commit:
[FastFile] add optional path parameter to #parse method

#parse method now has optional parameter path to parsing file, this is helpful when printing backtrace to console.
I also found problem with binding, some testing FastFile uses this new variable (actually there was method with same name,
but local variable has higher priority in this case) and causes to rewrite whole file.
So I add method #get_binding to resolve this issue.

Commit:
[FastFile] rename method #get_binding to #parsing_binding, because Rubocop doesn't like get_ prefix for methods

samnung commented:
Hi @KrauseFx, this PR is ready to be merged, could you look at it?

KrauseFx commented:
Haha, give me some time, I'm not online 24/7 :wink:

KrauseFx commented on lib/fastlane/fast_file.rb:
```diff
@@ -24,19 +24,38 @@ def initialize(path = nil)
+          if path.nil?
+            eval(data, parsing_binding) # this is okay in this case
+          else
+            eval(data, parsing_binding, relative_path) # this is okay in this case
```
Is the comment ` # this is okay in this case` correct here?

samnung replied:
I just duplicate the line and add parameters :smile:   so should I remove it? I think it is mentioning usage of eval function...

KrauseFx replied:
Getting rid of an `if else` is definitely worth a lot. Adding a comment there to explain the parameter is fine

KrauseFx commented on lib/fastlane/fast_file.rb:
```diff
@@ -24,19 +24,38 @@ def initialize(path = nil)
+
           # rubocop:disable Lint/Eval
-          eval(data) # this is okay in this case
+          if path.nil?
```
Do we even need the `if else` here? `relative_path` is `nil` anyway when the `path` is `nil`

samnung replied:
We have to, because it is trying to convert nil to String otherwise. I guess it detects number of arguments passing to `eval` function. I can change this to something like this:

```ruby
params = [data, parsing_binding, relative_path].compact
eval(*params)
```

Is this better?

KrauseFx commented on lib/fastlane/fast_file.rb:
```diff
@@ -24,19 +24,38 @@ def initialize(path = nil)
           # rubocop:enable Lint/Eval
         rescue SyntaxError => ex
-          line = ex.to_s.match(/\(eval\):(\d+)/)[1]
+          match = ex.to_s.match(/#{Regexp.escape(relative_path)}:(\d+)/) unless relative_path.nil?
```
Could you add one or two tests for this to make sure it works for different kinds of situations?

samnung replied:
Will do

KrauseFx commented:
Thanks for the pull request - it looks great. I added some comments before it's ready to be merged :+1:

Commit:
[FastFile] simplify code in #parse method

Commit:
[FastFile] add spec for printing correct message about syntax error in Fastfile

samnung commented:
Simplified code for `eval` and added two more tests, found some pretty bugs (it was using path from the binding, so it prints path to fast_file.rb :smile:). Fixed and I think it is ready to merge :blush:

KrauseFx commented:
Looks great. Yah, that's one of the main reasons for having tests, we can find those kinds of things. Thanks for your contribution, that's a great change in `fastlane` :+1:

samnung commented:
Yeah, I should write more tests... Glad to help :wink:
